src/core/manager.c

```diff
@@ -3306,6 +3315,80 @@ static int manager_deserialize_units(Manager *m, FILE *f, FDSet *fds) {
         return 0;
 }
 
+usec_t manager_get_runtime_watchdog(Manager *m) {
+        assert(m);
+
+        if (MANAGER_IS_USER(m))
+                return USEC_INFINITY;
+
+        if (timestamp_is_set(m->runtime_watchdog_overridden))
+                return m->runtime_watchdog_overridden;
+
+        return m->runtime_watchdog;
+}
+
+usec_t manager_get_reboot_watchdog(Manager *m) {
+        assert(m);
+
+        if (timestamp_is_set(m->reboot_watchdog_overridden))
+                return m->reboot_watchdog_overridden;
+
+        return m->reboot_watchdog;
+}
+
+usec_t manager_get_kexec_watchdog(Manager *m) {
+        assert(m);
+
+        if (timestamp_is_set(m->kexec_watchdog_overridden))
+                return m->kexec_watchdog_overridden;
+
+        return m->kexec_watchdog;
+}
+
+void manager_set_runtime_watchdog(Manager *m, usec_t runtime_watchdog) {
+        int r = 0;
+
+        assert(m);
+
+        if (MANAGER_IS_USER(m))
+                return;
+
+        if (m->runtime_watchdog == runtime_watchdog)
+                return;
+
+        if (!timestamp_is_set(m->runtime_watchdog_overridden)) {
+                if (timestamp_is_set(runtime_watchdog))
+                        r = watchdog_set_timeout(&runtime_watchdog);
+                else
+                        watchdog_close(true);
+        }
+
+        if (r >= 0)
+                m->runtime_watchdog = runtime_watchdog;
+}
+
+int manager_set_runtime_watchdog_overridden(Manager *m, usec_t runtime_watchdog) {
+        int r = 0;
+
+        assert(m);
+
+        if (MANAGER_IS_USER(m))
+                return 0;
+
+        if (m->runtime_watchdog_overridden == runtime_watchdog)
+                return 0;
+
+        if (timestamp_is_set(runtime_watchdog))
+                r = watchdog_set_timeout(&runtime_watchdog);
+        else
+                watchdog_close(true);
+
+        if (r >= 0)
+                m->runtime_watchdog_overridden = runtime_watchdog;
```

Comment:
OK, so let's say I call this function with -1, and trigger `watchdog_close()`. Then subsequent call to `manager_get_kexec_watchdog` will return the *non*-overridden value, which might be anything.

I don't think we need all that complication. Maybe just simplify things and once overridden, just set the flag and don't allow going back?

Reply:
No, assuming that you meant `manager_get_runtime_watchdog` and not `manager_get_kexec_watchdog`,  this function only set the non-overridden value, ie the value that reflects the setting from PID1 configuration. And it can't be anything, it's value is the one defined in system.conf basically.

Now if you think that restoring the value (by calling the D-BUS method with an empty or invalid timestamp) is not useful, we could reject these values but I would still keep the 2 timestamp values for the reason given in my previous comment.

Reply:
Right, I meant `manager_get_runtime_watchdog`, sorry.

After the call as described above, the watchdog if off (because we called `watchdog_close()`.). But a call to `manager_get_runtime_watchdog` will likely return some value that is not "off". (It *may* be correct only by lucky accident.)

Reply:
Oh I see now... good point. I'll fix it according to the outcome of your other concern.

Reply:
OK for now on, I'll restore the overridden value (if any) and only close the watchdog if this value is invalid.